text/0023-starlark_load_dataset.md

```diff
@@ -0,0 +1,249 @@
+
+### Relative "me" references aren't allowed:
+Any valid dataset reference should work with one exeption: no "me" statements.
+
+```python
+# this will error:
+pop = load_dataset("me/population")
+# this will work:
+pop = load_dataset("b5/population")
+```
+
+This is to keep transform scripts portable by freeing them from dependencies on execution context.
+
+This variable-style import makes sense for datasets, who's names are often lengthy. Unlike package imports, we anticipate it will be very common for dataset names to collide, being distinguished only be peername. this required assign-to-name pattern will help here as well.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+### differences with `load`
+As currently proposed users are expected to assign the result of `load_dataset`
+```python
+# load the `http` component of the http "package object"
+load("http.star", "http")
+# load a dataset into a variable named "fhv"
+fhv = load_dataset("b5/nyc_for_hire_vehicles")
+```
+
+In the `load` statement, the value `http` is defined by the module, which can only be known by reading the source. I find starlark's `load` statement opaque, and don't consider it a pattern we should follow.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+### load into a user-supplied string
+One alternative would be to have `load_dataset` work a little closer to the way `load` does in starlark, having a user supply a second string argument that names a global variable:
+```python
+# load b5/nyc_for_hire_vehicles into a global variable named "fhv"
+load_dataset("b5/nyc_for_hire_vehicles", "fhv")
+
+# use fhv:
+fhv.get_body()
+```
+
+I attempted to code this up, it was both difficult to execute, and felt too "magical" in practice. My initial reason for specing this syntax were due to the future need to make alterations that will make the return value behave in different ways during execution. consider the following scenario:
+```python
+# only allow access to "residents" during the "transform" step. "fhv" will 
+# be equal to None during all other steps
+load_dataset("b5/nyc_for_hire_vehicles", "fhv", steps=["transform"])
+
+def download(ctx):
+  fhv.get_body() # error: cannot access 'fhv' during download step
+
+def transform(ds,ctx):
+  fhv.get_body() # this will work
+```
+
+I now realize the problem is not in how the variable is declared, it is that the value stored at `fhv` is _stateful_. How the dataset is named is an orthagonal problem. The return-value style wins out in my mind on account of reading more clearly.
+
```

Comment:
I'm having a hard time understanding what this section is trying to express. 

Regardless of which import syntax we use, wouldn't the variable `fhv` have the same restrictions: `fhv.body()` would be able to be used in the transform step, but not the download step? 

Are you saying that by using a return-value syntax, we make it more clear that there is an object floating around that has special permissions/restrictions? Where the `load_dataset()` format doesn't make this as clear?

Reply:
Yeah I should re-work this section to remove the intermediate thinking ("I now realize the problem is...")

You have the point right exactly, the import syntax _has no bearing_ on the behavior of the returned value, and that's the only I'm trying to make.

 Here it's getting confused with a second idea: that `fhv` can behave differently during different phases of a transform. I describe this as being _stateful_, but that's a separate point. from the point I'm trying to make. 

I think the answer is to remove this section altogether, and instead just point out that the returned `fhv` value will be stateful, and it's behaviour will be dictated by the `dataset` object, which will be defined in a separate RFC.